content/en/blog/2021/wasm-api-alpha/index.md

```diff
@@ -0,0 +1,120 @@
+
+In addition, Istio 1.12 expands this capability to the Wasm OCI images. That means the Istio agent is now able to fetch Wasm images from any OCI registry including Docker Hub, Google Container Registry(GCR), Amazon Elastic Container Registry (Amazon ECR), etc. After fetching images, Istio-agent extracts and caches Wasm binaries from them, and then inserts them into Envoy filter chains.
+
+{{< image width="75%"
+    link="./istio-agent-architecture.svg"
+    alt="Remote Wasm module fetch flow"
+    caption="Remote Wasm module fetch flow"
+    >}}
+
+## Improvements in Envoy Wasm runtime
+
+The Wasm runtime powered by V8 in Envoy has been shipped as of Istio 1.5, and there has been a lot of improvements since then.
+
+### WASI supports
+
+First of all, some of the WASI (WebAssembly System Interface) system calls are supported now. For example, `clock_time_get` system call can be made from Wasm programs, which means that you can use `std::time::SystemTime::now()` in Rust or `time.Now().UnixNano()` in Go for Envoy Wasm extensions just like any other native platform. Another example is that `random_get` is supported by Envoy now, so the "crypto/rand" package is available in Go as a cryptographically secure random number generator. We are currently actively looking into file system support as we have seen the requests for reading and writing local files from Wasm programs running in Envoy.
+
+### Debuggability
+
+Next is the improvement in debuggability. Now the Envoy runtime emits the stack trace of your program when it causes runtime errors, for example, when null pointer exceptions occur in C++, or the panic function is called in Go or Rust. Previously Envoy did not say anything about the cause of errors, but now you can see the trace which can be used to debug your programs:
+
+{{< text plain >}}
+Function: proxy_on_request_headers failed: Uncaught RuntimeError: unreachable
+Proxy-Wasm plugin in-VM backtrace:
+  0:  0xdbd - runtime._panic
+  1:  0x103ab - main.anotherCalculation
+  2:  0x10399 - main.someCalculation
+  3:  0xea57 - main.myHeaderHandler
+  4:  0xea15 - proxy_on_request_headers
+{{< /text >}}
+
+The above is an example output of the stack trace for Go SDK based Wasm extensions. You might notice that you want file names and lines in the trace lines -- that is another huge future work and is currently an open issue since it has something to do with the DWARF format for WebAssembly and the Exception Handling proposal to WebAssembly specification.
+
+### Strace support for Wasm programs
+
+Also you can see strace equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, we can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example log stream of such strace logs:
```

Comment:
Curious: Is this something new added since 1.9? I thought we can already track all ABI calls before.

Reply:
I think it was between 1.7-1.8, but this section is about the improvements since 1.5 so I think it is fine to explain.